pkg/channel/event_receiver.go

```diff
@@ -0,0 +1,179 @@
+	"go.uber.org/zap"
+	"knative.dev/eventing/pkg/kncloudevents"
+	"knative.dev/eventing/pkg/utils"
+)
+
+var (
+	shutdownTimeout = 1 * time.Minute
+
+	// ErrUnknownChannel is returned when an event is received by a channel dispatcher for a
+	// channel that does not exist.
+	ErrUnknownChannel = errors.New("unknown channel")
+)
+
+// EventReceiver starts a server to receive new events for the channel dispatcher. The new
+// event is emitted via the receiver function.
+type EventReceiver struct {
+	ceClient          cloudevents.Client
+	receiverFunc      ReceiverFunc
+	logger            *zap.Logger
+	hostToChannelFunc ResolveChannelFromHostFunc
+}
+
+// ReceiverFunc is the function to be called for handling the event.
+type ReceiverFunc func(context.Context, ChannelReference, cloudevents.Event) error
+
+// ReceiverOptions provides functional options to EventReceiver function.
+type ReceiverOptions func(*EventReceiver) error
+
+// ResolveChannelFromHostFunc function enables EventReceiver to get the Channel Reference from incoming request HostHeader
+// before calling receiverFunc.
+type ResolveChannelFromHostFunc func(string) (ChannelReference, error)
+
+// ResolveChannelFromHostHeader is a ReceiverOption for NewEventReceiver which enables the caller to overwrite the
+// default behaviour defined by ParseChannel function.
+func ResolveChannelFromHostHeader(hostToChannelFunc ResolveChannelFromHostFunc) ReceiverOptions {
+	return func(r *EventReceiver) error {
+		r.hostToChannelFunc = hostToChannelFunc
+		return nil
+	}
+}
+
+// NewEventReceiver creates an event receiver passing new events to the
+// receiverFunc.
+func NewEventReceiver(receiverFunc ReceiverFunc, logger *zap.Logger, opts ...ReceiverOptions) (*EventReceiver, error) {
+	ceClient, err := kncloudevents.NewDefaultClient()
+	if err != nil {
+		logger.Fatal("failed to create cloudevents client", zap.Error(err))
+	}
+	receiver := &EventReceiver{
+		ceClient:          ceClient,
+		receiverFunc:      receiverFunc,
+		hostToChannelFunc: ResolveChannelFromHostFunc(ParseChannel),
+		logger:            logger,
+	}
+	for _, opt := range opts {
+		if err := opt(receiver); err != nil {
+			return nil, err
+		}
+	}
+	return receiver, nil
+}
+
+// Start begins to receive events for the receiver.
+//
+// Only HTTP POST requests to the root path (/) are accepted. If other paths or
+// methods are needed, use the HandleRequest method directly with another HTTP
+// server.
+func (r *EventReceiver) Start(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	errCh := make(chan error, 1)
+	go func() {
+		errCh <- r.ceClient.StartReceiver(ctx, r.receiverFunc)
+	}()
+
+	// Stop either if the receiver stops (sending to errCh) or if the context Done channel is closed.
+	select {
+	case err := <-errCh:
+		return err
+	case <-ctx.Done():
+		break
+	}
+
+	// Done channel has been closed, we need to gracefully shutdown h.ceClient. cancel() will start its
```

Comment:
```suggestion
	// Done channel has been closed, we need to gracefully shutdown h.ceClient.cancel() will start its
```

Reply:
Changed the comment as it was a bit misleading...